cmd/promtool/unittest.go

```diff
@@ -0,0 +1,467 @@
+							})
+						}
+					}
+
+					got[ar.Name()] = append(got[ar.Name()], alerts...)
+				}
+			}
+
+			for _, testcase := range alertTests[t] {
+				// Checking alerts.
+				gotAlerts := got[testcase.Alertname]
+
+				var expAlerts labelsAndAnnotations
+				for _, a := range testcase.ExpAlerts {
+					// User gives only the labels from alerting rule, which doesn't
+					// include this label (added by Prometheus during Eval).
+					a.ExpLabels[labels.AlertName] = testcase.Alertname
+
+					expAlerts = append(expAlerts, labelAndAnnotation{
+						Labels:      labels.FromMap(a.ExpLabels),
+						Annotations: labels.FromMap(a.ExpAnnotations),
+					})
+				}
+
+				if gotAlerts.Len() != expAlerts.Len() {
+					errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+						testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+				} else {
+					sort.Sort(gotAlerts)
+					sort.Sort(expAlerts)
+
+					if !reflect.DeepEqual(expAlerts, gotAlerts) {
+						errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+							testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+					}
+				}
+			}
+
+			curr++
+		}
+	}
+
+	// Checking promql expressions.
+	for _, testCase := range tg.PromqlExprTests {
+		got, err := query(suite.Context(), testCase.Expr, mint.Add(testCase.EvalTime),
+			suite.QueryEngine(), suite.Queryable())
+		if err != nil {
+			errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+				testCase.EvalTime.String(), err.Error()))
+		}
+
+		var gotSamples []parsedSample
+		for _, s := range got {
+			gotSamples = append(gotSamples, parsedSample{
+				Labels: s.Metric.Copy(),
+				Value:  s.V,
+			})
+		}
+
+		var expSamples []parsedSample
+		parseError := false
+		for _, s := range testCase.ExpSamples {
+			lb, err := promql.ParseMetric(s.Labels)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+					testCase.EvalTime.String(), err.Error()))
+				parseError = true
+				break
+			}
+			expSamples = append(expSamples, parsedSample{
+				Labels: lb,
+				Value:  s.Value,
+			})
+		}
+
+		if !parseError && !reflect.DeepEqual(expSamples, gotSamples) {
+			errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, \n        exp:%#v, \n        got:%#v", testCase.Expr,
+				testCase.EvalTime.String(), parsedSamplesString(expSamples), parsedSamplesString(gotSamples)))
+		}
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// seriesLoadingString returns the input series in
+// PromQL notation.
+func (tg *testGroup) seriesLoadingString() string {
+	result := ""
+	result += "load " + shortDur(tg.Interval) + "\n"
+	for _, is := range tg.InputSeries {
+		result += "  " + is.Series + " " + is.Values + "\n"
+	}
+	return result
+}
+
+// https://stackoverflow.com/a/41336257/6219247
```

Comment:
Do you have rights to use this code as apache2?

Reply:
I am unaware of licenses, but what I see is, if I remove that comment, the code is general enough that it can be used. Maybe change it a bit you say?